src/System.Runtime/src/System/HashCode.cs

```diff
@@ -0,0 +1,406 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    public struct HashCode
+    {
+#   if SYSTEM_HASHCODE_TESTVECTORS
+        private static readonly uint _seed = 0;
+#   else
+        private static readonly uint _seed = unchecked((uint)new Random().Next(int.MinValue, int.MaxValue));
```

Comment:
System.Security is not available in this assembly :).

Reply:
@bartonjs, can you please help with the incantations for getting proper entropy here?

Reply:
Initial thought: If seeded Marvin is enabled by default, string.Empty.GetHashCode() is entropic.  But there's probably a way of getting at randomness from within System.Runtime if I let myself think for more than 20 seconds.

Reply:
`new Random().Next(int.MinValue, int.MaxValue)` should be good enough source of entropy in .NET Core.

> `RandomNumberGenerator.Create().GetBytes()`

At this level of the stack, you can use SystemNative_GetNonCryptographicallySecureRandomBytes` and Windows equivalent if you have a good reason to - look for `Interop.GetRandomBytes` in CoreLib. The `Random` global seed is initialized by this function among other things.

Reply:
While you're at it beware that `static readonly` can lead to performance issues in this kind of code. Not only that it results in a call to ensure that the static constructor has ran but it also blocks constant folding that could otherwise be performed in `Initialize`. If performance is deemed important then you may need to look for a more involved solution (e.g. making _seed a field intrinsic (and of course, moving it somewhere in corelib) and having the VM initialize it with an good random value).

Reply:
@mikedn I tried to search for "field intrinsic" et al., but failed to encounter anything helpful. Could you give me a pointer (doc, example) on how to make such things work? For very small messages, constant-folding the initialization might have a non-negligible effect.

Reply:
The JIT has special handling for a few static readonly fields - e.g. `IntPtr.Zero`. It replaces a load from such a field with the actual constant without bothering to ensure that the static constructor has ran. Recently something like this was done for `BitConverter.IsLittelEndian` in https://github.com/dotnet/coreclr/pull/9789.

Unfortunately I now realize that this won't probably help with random seeds. The seed has to be different on every run so a constant value cannot be placed in a crossgened image. And it's exactly the crossgen scenario where this matters the most because then the static constructor calls cannot be elided.

Reply:
See https://github.com/dotnet/coreclr/issues/14592 and my first post in that issue for an example of the impact `static readonly` can have, the code is significantly larger when the JIT can't elide the static constructor check. But that hash function is much simpler and this the impact is probably far greater than in this case.

Reply:
Thanks for the examples!

Reply:
[I've reduced the number of reads of `s_seed` to one in Initialize.](https://github.com/jcdickinson/corefx/commit/b6ccf6940a6b8d5ed5cceee51cff0e13015f9aca#diff-6803f9310d27bb20dca3838086424542R260) Not sure if the jitter would have done this anyway with `readonly`, but it doesn't hurt to be cautious - this type is probably going to be a common hot path.

Reply:
The number of reads doesn't matter that much (or at all, I haven't checked the generated code).

And as always, it's best to measure before attempting optimizations. Try it with a fixed (const) seed and compare the result with the current version. If the difference is significant then we can consider alternatives. One possible solution is to avoid static constructors and instead use the good old fashioned `if (s_seed == 0) s_seed = ...;`

Intrinsic fields might still be an option but to use that we need to ensure that this code is never crossgened, that sounds kind of spooky.

Reply:
@jkotas , I think it's somewhat important to use a cryptographic random number generator here for a couple reasons:
1. We may change the algorithm in the future to account for DoS resistance, which would make a strong RNG critical
2. Non-cryptographic RNG can have seeding issues such that multiple AppDomains/processes/machines end up with the same seed.

Reply:
Having dependency from the core runtime parts on the full cryptographic stack would create ugly dependencies. I do think it is the right tradeoff between complexity and security. We had this discussion multiple times before, e.g. https://github.com/dotnet/corefx/pull/16652.

> We may change the algorithm in the future to account for DoS resistance

If we ever do this, we can re-evaluate what makes sense.

> Non-cryptographic RNG can have seeding issues 

If there are seeding issues in `SystemNative_GetNonCryptographicallySecureRandomBytes`, we should fix them.

Reply:
There is no real difference between any approach in terms of performance:

 |                  Method |     Mean |     Error |    StdDev | Scaled |
 |------------------------ |---------:|----------:|----------:|-------:|
 |           [Seed Read 1](https://github.com/jcdickinson/Benchmarks/blob/corefx-hashcode/ConsoleApp6/SeedRead1/HashCode.cs#L261) | 12.12 ns | 0.1618 ns | 0.1514 ns |   0.22 |
 |           [Seed Read 4](https://github.com/jcdickinson/Benchmarks/blob/corefx-hashcode/ConsoleApp6/SeedRead4/HashCode.cs#L261) | 12.01 ns | 0.1978 ns | 0.1850 ns |   0.21 |
 |             [Seed Lazy](https://github.com/jcdickinson/Benchmarks/blob/corefx-hashcode/ConsoleApp6/SeedLazy/HashCode.cs#L261) | 11.91 ns | 0.0165 ns | 0.0154 ns |   0.21 |
 |      [Seed Thread Lazy](https://github.com/jcdickinson/Benchmarks/blob/corefx-hashcode/ConsoleApp6/SeedThreadLazy/HashCode.cs#L261) | 13.85 ns | 0.0334 ns | 0.0261 ns |   0.25 |

Is there another place where I can see how the secure RNG is used so that I can chamber a change containing it?

Reply:
After further investigation with @bartonjs, CoreLib's Interop.GetRandomBytes is acceptable on most machines, so I can live with it here. @jcdickinson, I think the best way to get that would be to actually move this change to System.Private.CoreLib in the https://github.com/dotnet/coreclr/ repo after we sign off here (we'll just merge as soon as CI goes green assuming nothing meaningful changes in the move). Make sure to put it in the src/mscorlib/shared directory so that it automatically mirrors to CoreRT. That'll give you access to the API and also let CoreLib use HashCode. 

@jkotas, for reference, the issues in GetNonCryptographicallySecureRandomBytes are:
1. It can use urandom before it's seeded, meaning it may be predictable or even constant between containers/VMs/IoT devices where entropy trickles in slowly and a .NET service may run on boot.
2. It silently fails over from random/urandom to srand(time(null)) if a machine doesn't support them or they're too slow. If this happens, the seed is both very predictable and could easily be the same across processes or machines. It also means cases like leap seconds and VM rollbacks will cause repeat behavior.